File: day_7/puzzle_1/main.py
from statistics import median
from functools import reduce

def read_input():
    with open("input.txt", "r") as input_file:
        return list(map(int, input_file.readline().split(",")))

data = read_input()

final_position = int(median(data))
print(reduce(lambda x,y: x + abs(y-final_position), data, 0))

File: day_7/puzzle_1/main.hs
import Data.List.Split (splitOn)
import Data.List (sort)

main :: IO ()
main = do
  userInput <- getContents
  print $ floor . getFuelUsed $ mapInput userInput

mapInput :: Read b => [Char] -> [b]
mapInput xs = map read (splitOn "," xs)

getMedian :: (Ord p, Fractional p) => [p] -> p
getMedian [] = error "Empty list for median"
getMedian xs
  | odd len = xs !! mid
  | even len = evenMedian
    where
      sorted_xs = sort xs
      len = length xs
      mid = len `div` 2
      evenMedian = (sorted_xs !! (mid-1) + sorted_xs !! mid) / 2
getMedian _ = error "This shouldn't be here"

getFuelUsed :: (Ord a, Fractional a) => [a] -> a
getFuelUsed xs = foldl (\x y -> x + abs(y-finalPosition)) 0 xs
  where finalPosition = getMedian xs
